src/aio/native_linux_aio_provider.cpp

```diff
@@ -26,107 +26,119 @@
+            return nullptr;
+        }
     }
 
-    LOG_ERROR("close file failed, err = {}", utils::safe_strerror(errno));
-    return ERR_FILE_OPERATION_FAILED;
+    // Open the file for write as RandomRWFile, to support un-sequential write.
+    std::unique_ptr<rocksdb::RandomRWFile> wfile;
+    s = dsn::utils::PegasusEnv(dsn::utils::FileDataType::kSensitive)
+            ->NewRandomRWFile(fname, &wfile, rocksdb::EnvOptions());
+    if (!s.ok()) {
+        LOG_ERROR("open write file '{}' failed, err = {}", fname, s.ToString());
+    }
+    return wfile;
 }
 
-error_code native_linux_aio_provider::flush(linux_fd_t fd)
+error_code native_linux_aio_provider::close(rocksdb::RandomRWFile *wf)
 {
-    if (fd.is_invalid() || ::fsync(fd.fd) == 0) {
-        return ERR_OK;
+    auto s = wf->Close();
+    if (!s.ok()) {
+        LOG_ERROR("close file failed, err = {}", s.ToString());
+        return ERR_FILE_OPERATION_FAILED;
     }
 
-    LOG_ERROR("flush file failed, err = {}", utils::safe_strerror(errno));
-    return ERR_FILE_OPERATION_FAILED;
+    return ERR_OK;
+}
+
+error_code native_linux_aio_provider::flush(rocksdb::RandomRWFile *wf)
+{
+    auto s = wf->Fsync();
+    if (!s.ok()) {
+        LOG_ERROR("flush file failed, err = {}", s.ToString());
+        return ERR_FILE_OPERATION_FAILED;
+    }
+
+    return ERR_OK;
 }
 
 error_code native_linux_aio_provider::write(const aio_context &aio_ctx,
                                             /*out*/ uint64_t *processed_bytes)
 {
-    dsn::error_code resp = ERR_OK;
-    uint64_t buffer_offset = 0;
-    do {
-        // ret is the written data size
-        auto ret = ::pwrite(aio_ctx.dfile->native_handle().fd,
-                            (char *)aio_ctx.buffer + buffer_offset,
-                            aio_ctx.buffer_size - buffer_offset,
-                            aio_ctx.file_offset + buffer_offset);
-        if (dsn_unlikely(ret < 0)) {
-            if (errno == EINTR) {
-                LOG_WARNING("write failed with errno={} and will retry it.",
-                            utils::safe_strerror(errno));
-                continue;
-            }
-            resp = ERR_FILE_OPERATION_FAILED;
-            LOG_ERROR("write failed with errno={}, return {}.", utils::safe_strerror(errno), resp);
-            return resp;
-        }
-
-        buffer_offset += ret;
-        if (dsn_unlikely(buffer_offset != aio_ctx.buffer_size)) {
-            LOG_WARNING(
-                "write incomplete, request_size={}, total_write_size={}, this_write_size={}, "
-                "and will retry it.",
-                aio_ctx.buffer_size,
-                buffer_offset,
-                ret);
-        }
-    } while (dsn_unlikely(buffer_offset < aio_ctx.buffer_size));
+    rocksdb::Slice data((const char *)(aio_ctx.buffer), aio_ctx.buffer_size);
+    auto s = aio_ctx.dfile->wfile()->Write(aio_ctx.file_offset, data);
+    if (!s.ok()) {
+        LOG_ERROR("write file failed, err = {}", s.ToString());
```

Comment:
Is it necessary to specify which file fails to be written ?

Reply:
Not necessary, the callers have the responsibility to log the fikle name.